File: iitj_assignments/assignment_1/python programs/8_Circular queue.py
#!/usr/bin/env python
# coding: utf-8

# In[2]:


class node:
    #initiation
    def __init__(self, data):
        self.data = data
        self.next = None

class circular_Queue:
    #initiation
    def __init__(self):
        self.front = self.rear = None
    #queue is empty or not
    def isEmpty(self):
        return self.front == None

    def enqueue(self, data): #insert at rear end
        temp = node(data)
        if self.isEmpty():
            self.front = temp
        else:
            self.rear.next = temp
        self.rear = temp
        self.rear.next = self.front

    def dequeue(self):   #delete from front end
        if self.isEmpty():
            print("underflow\n")
            return
        print(self.front.data ," is deleted\n")
        if self.front==self.rear : #only one element left
            self.front = self.rear = None
        else:
            self.front = self.front.next
            self.rear.next = self.front
    #display queue
    def display(self):
        temp = q.front
        print("front  ",end=" ")
        while (temp.next != self.front):
            print(temp.data, end = " ")
            temp = temp.next
        print(temp.data,"    rear\n")

q=circular_Queue()
q.enqueue(1)
q.enqueue(2)
q.enqueue(3)
q.display()
q.dequeue()
q.display()
q.enqueue(4)
q.enqueue(5)
q.display()
q.dequeue()
q.display()

File: iitj_assignments/assignment_1/c programs/8_circular_queue.c
#include<stdio.h>
#include<stdlib.h>
#include <stdbool.h>
//circular queue node
struct circular_queue
{
  int rear;
  int front;
  int capacity;
  int *array;
};
//initialization
struct circular_queue* creation()
{
  struct circular_queue *q;
  q->capacity =100;
  q->front = q->rear = -1;
  q->array = (int *)malloc(q->capacity*sizeof(int));
  return q;
}
//queue is full or not
bool isFull(struct circular_queue *q)
{
  if(((q->rear+1)%q->capacity)==q->front)
    return true;
  return false;
}
//queue is empty or not
bool isEmpty(struct circular_queue *q)
{
  if(q->rear==-1 && q->front==-1)
    return true;
  return false;
}
//enqueue element at rear end
void enqueue(struct circular_queue *q,int data)
{
  if(isFull(q))
    printf("Overflow\n");
  if(q->front ==-1 && q->rear ==-1)
        q->front = q->rear=0;
  else
        q->rear = (q->rear+1)%q->capacity;
  q->array[q->rear]=data;
}
//dequeue element from front end
int dequeue(struct circular_queue *q)
{
  if(isEmpty(q))
    {
      printf("underflow\n");
      return -1;
    }
  int temp = q->array[q->front];
  if(q->front==q->rear)
     q->front=q->rear =-1;
  else
     q->front = (q->front+1)%q->capacity;

     return temp;
}
//main function
int main()
{
    int choice,data,temp;
    struct circular_queue *q=creation();
    while(1)
    {
        printf("*************************************\n");
        printf("1.enqueue data\n");
        printf("2.dequeue data\n");
        printf("3.Exit\n");
        printf("enter your choice \n");
        scanf("%d",&choice );
        switch(choice)
        {
            case 1:
                    printf("enter your data \n");
                    scanf("%d",&data );
                    enqueue(q,data);
                    break;
            case 2:
                    temp = dequeue(q);
                    if(temp!=-1)
                       printf("Dequed element is %d\n",temp);
                    break;
            case 3:
                    exit(0);
            default:
                    printf("wrong choice\n");
                    break;
        }
    }
}
